docs/design-patterns.md

```diff
@@ -0,0 +1,28 @@
+
+TL;DR; Datastore can no longer be magic singleton, bc complicates testing and various
+settings (namespace for now, potentially databaseId, host, and even project) cannot be
+shared across calls. SO this requires that we:
+
+1. modify all interfaces to take either `Datastore` client, `DatastoreOptions`, or some abstract 
+   `Options` class that incorporates the configuration information contained in those
+2. make all classes dependent on `Datastore` not singletons; and take it as a constructor parameter
+
+Current choice is (2), because limits scope of changes for users; will just have to change how they
+get `PipelineService`/`PipelineManager` instances for launching jobs; and initially, for case of 
+running everything within default project/database/namespace, it can remain a singleton for them.
+
+
+## What is the right pattern long-term?
+
+Passing options around allows classes to be self-evidently threadsafe; no risk that one thread
```

Comment:
everything happens in request/response model, right? Can we pass through something (a header, a param) that identifies the org and provide a Database instance in a filter, available for the whole request?